### DISCUSSION TITLE
Jackson deserialization failing

### DISCUSSION THREAD

**Discussion by rsenden** [micronaut-projects/micronaut-core#6393] [Q&A] [answered]
I have Micronaut inject its Jackson ObjectMapper into one of my singletons, and then call ObjectMapper#readValue(String, Class) to deserialize the class listed below. This works fine in a normal JVM, but in a GraalVM native image this results in the following exception:

```
Caused by: java.lang.UnsupportedOperationException: Cannot write a read-only property: basicConnectionConfig
        at io.micronaut.inject.beans.AbstractInitializableBeanIntrospection$BeanPropertyImpl.set(AbstractInitializableBeanIntrospection.java:435)
        at io.micronaut.jackson.modules.BeanIntrospectionModule$VirtualSetter.deserializeAndSet(BeanIntrospectionModule.java:610)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:324)
```

Any ideas what is causing this, and how it can be fixed?

```
@Data @Introspected @JsonIgnoreProperties(ignoreUnknown = true)
public class MyConnectionConfig implements IBasicConnectionConfigProvider {
	private BasicConnectionConfig basicConnectionConfig;
	private BasicUserCredentialsConfig basicUserCredentialsConfig;
	private boolean renewAllowed;
	private char[] token;
	
	@JsonIgnore public final AuthType getAuthType() {
		return token!=null && token.length>0 ? AuthType.TOKEN : AuthType.USER;
	}
	
	public static enum AuthType {
		TOKEN, USER
	}

	@JsonIgnore public final boolean hasUserCredentialsConfig() {
		return basicUserCredentialsConfig!=null 
				&& StringUtils.isNotEmpty(basicUserCredentialsConfig.getUser())
				&& basicUserCredentialsConfig.getPassword()!=null;
	}
}
```

Edit: In my original question these fields were declared final, based on my expectation that Jackson would follow the getBasicRestConnectionConfig() method to set the properties on the existing instance like some other frameworks would do. I have since found out that Jackson doesn't do this (except for Maps and Collections if `MapperFeature.USE_GETTERS_AS_SETTERS` is enabled), but I'm still getting the same exceptions if I declare these fields without `final` modifier and don't initialize them.

**Comment by rsenden** [comment, marked as answer]
Yet again answering my own question; the problem wasn't in this class but in another class that includes this class as a field. Basically the structure was like this:

```
Class1:
  private Class2 fieldForClass2;
  getFieldForClass3() { return fieldForClass2.getFieldForClass3; }

Class2:
  private Class3 fieldForClass3;
  get/setFieldForClass3
```

I was assuming that this error was related to Class2, as the exception seemed to reference the fieldForClass3 field/methods defined in that class. 

However, I forgot to add `@JsonIgnore` to Class1#getFieldForClass3(), so basically Jackson was serializing my Class3 twice; once as part of Class1 and once as part of Class2. When trying to deserialize, Jackson obviously failed to deserialize Class3 into Class1 as there is no field or setter in that class.

Question remains of course why behavior is different between native image and JVM-based run...

**Reply by yawkat** [reply]
Unfortunately, because of the lack of reflection information on native-image, there are some internal differences with how serialization properties are discovered. Also compare #6403 on this.
